updater/aws_test.go

```diff
@@ -15,6 +15,260 @@ import (
+			}, nil
+		},
+		SendCommandFn: func(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error) {
+			return &ssm.SendCommandOutput{
+				Command: &ssm.Command{
+					CommandId:    aws.String("command-id"),
+					DocumentName: aws.String("check-document"),
+				},
+			}, nil
+		},
+		WaitUntilCommandExecutedWithContextFn: func(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error {
+			assert.Equal(t, "command-id", aws.StringValue(input.CommandId))
+			return nil
+		},
+	}
+	u := updater{ssm: mockSSM, checkDocument: "check-document"}
+	actual, err := u.filterAvailableUpdates(instances)
+	require.NoError(t, err)
+	assert.Equal(t, expected, actual, "Should only contain instances in Aavailable or Ready update state")
+}
+
+func TestPaginatedFilterAvailableUpdatesSuccess(t *testing.T) {
+	checkPattern := "{\"update_state\": \"%s\", \"active_partition\": { \"image\": { \"version\": \"%s\"}}}"
+	expected := make([]instance, 0)
+	instances := make([]instance, 0)
+	getOut := &ssm.GetCommandInvocationOutput{
+		Status:                aws.String("Success"),
+		StandardOutputContent: aws.String(fmt.Sprintf(checkPattern, updateStateAvailable, "v1.0.5")),
+	}
+
+	for i := 0; i < 100; i++ {
+		containerID := "cont-inst-br" + strconv.Itoa(i)
+		ec2ID := "ec2-id-br" + strconv.Itoa(i)
+		instances = append(instances, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+		})
+		expected = append(expected, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+			bottlerocketVersion: "v1.0.5",
+		})
+	}
+
+	mockSSM := MockSSM{
+		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
+			return getOut, nil
+		},
+		SendCommandFn: func(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error) {
+			return &ssm.SendCommandOutput{
+				Command: &ssm.Command{
+					CommandId:    aws.String("command-id"),
+					DocumentName: aws.String("check-document"),
+				},
+			}, nil
+		},
+		WaitUntilCommandExecutedWithContextFn: func(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error {
+			assert.Equal(t, "command-id", aws.StringValue(input.CommandId))
+			return nil
+		},
+	}
+	u := updater{ssm: mockSSM}
+	actual, err := u.filterAvailableUpdates(instances)
+	require.NoError(t, err)
+	assert.EqualValues(t, expected, actual, "should contain all instances")
+}
+
+func TestPaginatedFilterAvailableUpdatesAllFail(t *testing.T) {
+	instances := make([]instance, 0)
+
+	for i := 0; i < 100; i++ {
+		containerID := "cont-inst-br" + strconv.Itoa(i)
+		ec2ID := "ec2-id-br" + strconv.Itoa(i)
+		instances = append(instances, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+		})
+	}
+
+	mockSSM := MockSSM{
+		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
+			return nil, nil
+		},
```

Comment:
Does this need to be defined?  If all calls to `SendCommand` fail, does `GetCommandInvocation` get invoked at all?

Reply:
No, it doesn't need to be defined. Good eye, I'll remove the definition.